libs/small-steps/src/Data/Sharing.hs

```diff
@@ -0,0 +1,194 @@
+    internsFromMap,
+    internsFromVMap,
+    toMemptyLens,
+  )
+where
+
+import Cardano.Binary (Decoder, FromCBOR (..), decodeListLen, dropMap)
+import Control.Iterate.SetAlgebra (BiMap (..), biMapFromMap)
+import Control.Monad (void)
+import Control.Monad.Trans
+import Control.Monad.Trans.State.Strict
+import Data.Coders (decodeMap, decodeVMap, invalidKey)
+import Data.Compact.VMap (VB, VMap, VP)
+import qualified Data.Compact.VMap as VMap
+import qualified Data.Foldable as F
+import Data.Kind
+import qualified Data.Map.Strict as Map
+import Data.Map.Strict.Internal
+import Data.Primitive.Types (Prim)
+import Lens.Micro
+
+-- =======================================
+
+data Intern a = Intern
+  { internMaybe :: a -> Maybe a,
+    -- | Used for sorting
+    internWeight :: Int
+  }
+
+newtype Interns a = Interns [Intern a]
+  deriving (Monoid)
+
+interns :: Interns k -> k -> k
+interns (Interns []) !k = k -- opimize for common case when there are no interns
+interns (Interns is) !k = go is
+  where
+    go [] = k
+    go (x : xs) =
+      case internMaybe x k of
+        Just kx -> kx
+        Nothing -> go xs
+{-# INLINE interns #-}
+
+internsFromMap :: Ord k => Map k a -> Interns k
+internsFromMap m =
+  Interns
+    [ Intern
+        { internMaybe = \k ->
+            let go Tip = Nothing
+                go (Bin _ kx _ l r) =
+                  case compare k kx of
+                    LT -> go l
+                    GT -> go r
+                    EQ -> Just kx
+             in go m,
+          internWeight = Map.size m
+        }
+    ]
+
+internsFromVMap :: Ord k => VMap VB kv k a -> Interns k
+internsFromVMap m =
+  Interns
+    [ Intern
+        { internMaybe = \k -> VMap.internMaybe k m,
+          internWeight = VMap.size m
+        }
+    ]
+
+instance Semigroup (Interns a) where
+  (<>) is1 (Interns []) = is1
+  (<>) (Interns []) is2 = is2
+  (<>) (Interns is1) (Interns is2) =
+    Interns (F.foldl' insertIntoSortedInterns is2 is1)
+    where
+      insertIntoSortedInterns [] i = [i]
+      insertIntoSortedInterns (a : as) i
+        | internWeight a > internWeight i = a : insertIntoSortedInterns as i
+        | otherwise = i : a : as
+
+class Monoid (Share a) => FromSharedCBOR a where
+  {-# MINIMAL ((getShare, fromSharedCBOR) | fromSharedPlusCBOR) #-}
+  type Share a :: Type
+  type Share a = ()
+
+  getShare :: a -> Maybe (Share a)
+  getShare _ = Nothing
+
+  -- | Utilize sharing when decoding, but do not add anything to the state for
+  -- future sharing.
+  fromSharedCBOR :: StateT (Share a) (Decoder s) a
+  fromSharedCBOR = do
+    s <- get
+    x <- fromSharedPlusCBOR
+    x <$ put s
+
+  -- | Deserialize with sharing and add the state that used for sharing. Default
+  -- implementation will add value returned by `getShare` for adding to the
+  -- state.
+  fromSharedPlusCBOR :: StateT (Share a) (Decoder s) a
+  fromSharedPlusCBOR = do
+    x <- fromSharedCBOR
+    x <$ mapM_ (\s -> modify' (s <>)) (getShare x)
```

Comment:
@TimSheard I took your suggestion and brought in the `fromSharedPlusCBOR` function into the  class. This allows us to implement decoding with sharing in two ways:

1. implementing `fromSharedCBOR` and `getShare` (example instances are in this module)
2. implementing only `fromSharedPlusCBOR`, in which case the state used for sharing needs to be modified manually and `getShare` is completely ignored.

`MINIMAL` pragma will warn the programmer if it wasn't done right